assembly/src/assembler/instruction/crypto_ops.rs

```diff
@@ -96,19 +96,58 @@ pub(super) fn mtree_get(span: &mut SpanBuilder) -> Result<Option<CodeBlock>, Ass
 ///
 /// This operation takes 14 VM cycles.
 pub(super) fn mtree_set(span: &mut SpanBuilder) -> Result<Option<CodeBlock>, AssemblyError> {
+    // input: [d, i, R_old, V_new, ...]
+
     // Inject the old node value onto the stack for the call to MRUPDATE.
-    // [d, i, R, V_new, ...] => [V_old, d, i, R, V_new, ...]
+    // stack: [V_old, d, i, R_old, V_new, ...]
     read_mtree_node(span);
+
+    // Note: The stack is 14 elements deep already. The maximum depth the VM can go is 16, so it's
+    // only possible to copy 2-elements at the time.
+    #[rustfmt::skip]
+    let copy_old_element_ops = [
+        // Renamed V_old => o<pos>, R_old => r<pos>, V_new => n<pos>
+        // stack: [[o0, o1, o2, o3], [d, i, r0, r1], [r2, r3, n0, n1], n2, n3, ...]
+
+        // Copy last two elements of the word, and move it down
+        // stack: [[o0, o1, o2, o3], [d, i, r0, r1], [o3, o2, r2, r3], [n0, n1, n2, n3], ...]
+        Dup2, MovDn8, Dup3, MovDn8,
+
+        // Fix the position of the new elements
+        // stack: [[o3, o2, r2, r3], [n0, n1, n2, n3], [o0, o1, o2, o3], [d, i, r0, r1], ...]
+        SwapDW,
+
+        // stack: [[r2, r3, n0, n1], [n2, n3, o2, o3], [o0, o1, o2, o3], [d, i, r0, r1], ...]
+        MovDn7, MovDn6,
+
+        // The stack is restored, but with the two elements copied at the end
+        // stack: [[o0, o1, o2, o3], [d, i, r0, r1], [r2, r3, n0, n1], [n2, n3, o2, o3], ...]
+        SwapDW,
+
+        // Repeat for the other two elements
+        Dup1, MovDn8, Dup0, MovDn8, SwapDW, MovDn7, MovDn6, SwapDW
+    ];
+
+    // stack: [V_old, d, i, R_old, V_new, V_old, ...]
+    span.add_ops(copy_old_element_ops)?;
+
+    // stack: [R_new, R_old, V_new, V_old, ...]
     update_mtree(span, false);
+
     #[rustfmt::skip]
-    let ops = [
-        // Move the old root to the top of the stack => [R, R_new, V_new, ...]
-        SwapW,
+    let remove_depth_and_index_ops = [
```

Comment:
Just curious about the choice of name of this variable, are we removing the depth and index?

Reply:
oops, my mistake, I was thinking of `update_mtree` behavior. Fixed :)